src/main/kotlin/com/devh/vitstore/common/error/ApiExceptionHandler.kt

```diff
@@ -2,33 +2,322 @@ package com.devh.vitstore.common.error
-// import org.springframework.validation.FieldError
-// import org.springframework.validation.ObjectError
-//
-// import java.util.HashMap
-//
-// import org.springframework.web.bind.MethodArgumentNotValidException
-// import org.springframework.web.bind.annotation.ExceptionHandler
-// import org.springframework.web.bind.annotation.ResponseStatus
-// import java.util.function.Consumer
-
-// @ControllerAdvice
-// class ApiExceptionHandler : ResponseEntityExceptionHandler() {
-//    override fun handleHttpMessageNotReadable(ex: HttpMessageNotReadableException, headers: HttpHeaders, status: HttpStatus, request: WebRequest): ResponseEntity<Any> {
-//        val apiError = ApiErrorRes(status, ApiErrorDto(123, ex.localizedMessage))
-//        return ResponseEntity(apiError, HttpHeaders(), apiError.status)
-//    }
-//
-//    override fun handleNoHandlerFoundException(ex: NoHandlerFoundException, headers: HttpHeaders, status: HttpStatus, request: WebRequest): ResponseEntity<Any> {
-//        val apiError = ApiErrorRes(HttpStatus.NOT_FOUND, arrayListOf(ApiErrorDto(456, ex.localizedMessage), ApiErrorDto(789, "No handler found for ")))
-//        return ResponseEntity(apiError, HttpHeaders(), apiError.status)
-//    }
-// }
+import com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException
+import org.springframework.beans.ConversionNotSupportedException
+import org.springframework.beans.TypeMismatchException
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.http.converter.HttpMessageNotReadableException
+import org.springframework.http.converter.HttpMessageNotWritableException
+import org.springframework.web.HttpMediaTypeNotAcceptableException
+import org.springframework.web.HttpMediaTypeNotSupportedException
+import org.springframework.web.HttpRequestMethodNotSupportedException
+import org.springframework.web.bind.*
+import org.springframework.web.bind.annotation.RestControllerAdvice
+import org.springframework.web.context.request.WebRequest
+import org.springframework.web.context.request.async.AsyncRequestTimeoutException
+import org.springframework.web.multipart.support.MissingServletRequestPartException
+import org.springframework.web.servlet.NoHandlerFoundException
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
+import java.text.MessageFormat
+import java.util.stream.Collectors
+
+@RestControllerAdvice
+class ApiExceptionHandler : ResponseEntityExceptionHandler() {
+    override fun handleHttpRequestMethodNotSupported(
+        ex: HttpRequestMethodNotSupportedException,
+        headers: HttpHeaders,
+        status: HttpStatus,
+        request: WebRequest
+    ): ResponseEntity<Any> {
+
+        pageNotFoundLogger.warn(ex.message)
+
+        val supportedMethods = ex.supportedHttpMethods
+        if (supportedMethods!!.isNotEmpty()) {
```

Comment:
E sửa với not-null nhé, ko sử dụng `!!` trong các trường hợp chưa chắc chắn (bọn e có thể quy định ko dùng trong mọi trường hợp cũng được) 
```
if (supportedMethods?.isNotEmpty() == true) {
}
```